### PR TITLE
Ban "write" access for fields on continuous function spaces (for #138)

### PR DESCRIPTION
**Pull request by TeranIvy**
This PR removes `GH_WRITE` from the list of allowed accesses for fields on continuous function spaces. `ANY_SPACE` access for different looping types is dealt with in #471.

There are many test changes due to several test kernels using this access for continuous spaces. Some tranformation tests for parallel OMP DO loops were modified to use test examples on discontinuous spaces as it is not possible to do OMP on continuous spaces without colouring. Some of the modified test kernels had Fortran that could not be compiled so the code was corrected with generating subroutine body and enabling compilation in relevant tests.

### CODE REVIEW

**Comment by TeranIvy** [comment]
Ready for review from @rupertford.

File: src/psyclone/tests/dynamo0p3_test.py [hunk 1]
```diff
@@ -5739,50 +5750,49 @@ def test_unexpected_type_error(dist_mem):
...
-                       create_arg_list.diff_basis,
-                       create_arg_list.orientation,
-                       create_arg_list.field_bcs_kernel,
-                       create_arg_list.operator_bcs_kernel,
-                       create_arg_list.banded_dofmap,
-                       create_arg_list.indirection_dofmap,
-                       create_arg_list.cma_operator]:
-            with pytest.raises(NotImplementedError):
-                method(None)
+def test_argordering_exceptions(dist_mem):
```
Comment (rupertford) [inline src/psyclone/tests/dynamo0p3_test.py:5753]: Great again.

File: src/psyclone/tests/dynamo0p3_test.py [hunk 2]
```diff
@@ -5990,25 +6000,27 @@ def test_anyw2_operators(dist_mem, tmpdir):
...
-            "(STENCIL_CROSS,extent)\n"
-            "      f2_stencil_dofmap => f2_stencil_map%get_whole_dofmap()\n"
-            "      f2_stencil_size = f2_stencil_map%get_size()\n"
-            "      !\n")
-        assert output in generated_code
+    psy = PSyFactory(TEST_API,
+                     distributed_memory=dist_mem).create(invoke_info)
+    generated_code = str(psy.gen)
+
+    assert LFRicBuild(tmpdir).code_compiles(psy)
```
Comment (rupertford) [inline src/psyclone/tests/dynamo0p3_test.py:6013]: Brill.

File: src/psyclone/tests/dynamo0p3_test.py [hunk 3]
```diff
@@ -6635,12 +6653,14 @@ def test_no_halo_exchange_annex_dofs(tmpdir, monkeypatch,
...
     result = str(psy.gen)
+
     assert LFRicBuild(tmpdir).code_compiles(psy)
+
     if annexed:
         assert "CALL f1_proxy%halo_exchange" not in result
     else:
         assert "CALL f1_proxy%halo_exchange" in result
-    assert "CALL f2_proxy%halo_exchange" not in result
+        assert "CALL f2_proxy%halo_exchange" in result
```
Comment (rupertford) [inline src/psyclone/tests/dynamo0p3_test.py:6663]: Please add a check in `if annexed` that the call does not exist in the output.

File: src/psyclone/tests/test_files/dynamo0p3/loop_fuse_trans.py [hunk 4]
```diff
@@ -1,19 +1,56 @@
+# -----------------------------------------------------------------------------
```
Comment (rupertford) [inline src/psyclone/tests/test_files/dynamo0p3/loop_fuse_trans.py:1]: Thanks for adding the missing copyright notice.

File: src/psyclone/tests/test_files/dynamo0p3/testkern_anyw2_stencil_mod.f90 [hunk 5]
```diff
@@ -24,21 +28,49 @@
...
+     integer :: iterates_over = cells
    contains
-     procedure() :: code => testkern_anyw2_stencil_code
+     procedure, nopass :: code => testkern_anyw2_stencil_code
   end type testkern_anyw2_stencil_type
+
 contains
-  subroutine testkern_anyw2_stencil_code()
+
+  subroutine testkern_anyw2_stencil_code(nlayers, fld1,                 &
```
Comment (rupertford) [inline src/psyclone/tests/test_files/dynamo0p3/testkern_anyw2_stencil_mod.f90:56]: Thanks for adding the arguments.

File: src/psyclone/domain/lfric/lfric_arg_descriptor.py [hunk 6]
```diff
@@ -344,15 +344,13 @@ def _validate_field(self, arg_type):
...
-        # 'ANY_SPACE_*' will have the same allowed accesses as discontinuous
-        # fields as they do not need colouring when updating quantities
-        # in parallel.
-        field_cont_accesses = [AccessType.READ, AccessType.INC,
-                               AccessType.WRITE]
+        field_cont_accesses = [AccessType.READ, AccessType.INC]
+        # TODO in issue #471: The kernels that loop over DoFs (built-ins)
+        # and are specified on 'ANY_SPACE_*' will have the same allowed
+        # accesses as discontinuous fields as they do not need colouring
+        # when updating quantities in parallel. This temporary accesses
```
[the remark this answers is not in the document]
Comment (TeranIvy) [reply src/psyclone/domain/lfric/lfric_arg_descriptor.py:351]: Thanks for the suggestion, done.

File: src/psyclone/domain/lfric/lfric_arg_descriptor.py [hunk 7]
```diff
@@ -373,6 +371,10 @@ def _validate_field(self, arg_type):
         # Check fields on continuous function spaces
         fld_cont_acc_msg = [rev_access_mapping[acc] for acc in
                             field_cont_accesses]
+        # As said above, this temporary access for messages will be
```
[the remark this answers is not in the document]
Comment (TeranIvy) [reply src/psyclone/domain/lfric/lfric_arg_descriptor.py:374]: Done.

File: src/psyclone/domain/lfric/lfric_arg_descriptor.py [hunk 8]
```diff
@@ -384,14 +386,14 @@ def _validate_field(self, arg_type):
                        rev_access_mapping[self._access_type], arg_type))
         # As said above, allowed accesses for fields on ANY_SPACE depend on
         # type of looping. This will be refined in #471, however for now we
-        # use the accesses for continuous spaces (worst-case scenarios).
+        # use the temporary accesses for continuous spaces.
```
[the remark this answers is not in the document]
Comment (TeranIvy) [reply src/psyclone/domain/lfric/lfric_arg_descriptor.py:389]: Done.

File: src/psyclone/tests/domain/lfric/transformations/dynamo0p3_transformations_test.py [hunk 9]
```diff
@@ -613,12 +615,20 @@ def test_check_seq_colours_omp_do(tmpdir, monkeypatch, annexed, dist_mem):
...
-def test_colouring_after_openmp(dist_mem):
-    '''Test that we raise an error if the user attempts to colour a loop
+def test_colouring_after_openmp(dist_mem, monkeypatch):
+    ''' Test that we raise an error if the user attempts to colour a loop
     that is already within an OpenMP parallel region. We test when
-    distributed memory is on or off. '''
-    _, invoke = get_invoke("1_single_invoke.f90", TEST_API,
-                           name="invoke_0_testkern_type", dist_mem=dist_mem)
+    distributed memory is on or off.
+    Note: OpenMP parallel transformation can be applied on its own only on
```
[the remark this answers is not in the document]
Comment (TeranIvy) [reply src/psyclone/tests/domain/lfric/transformations/dynamo0p3_transformations_test.py:622]: Done.

File: src/psyclone/tests/domain/lfric/transformations/dynamo0p3_transformations_test.py [hunk 10]
```diff
@@ -613,12 +615,20 @@ def test_check_seq_colours_omp_do(tmpdir, monkeypatch, annexed, dist_mem):
...
+    a discontinuous function space. However, colouring can be applied when
+    an argument has "INC" access and in LFRic API this is only possible
```
[the remark this answers is not in the document]
Comment (TeranIvy) [reply src/psyclone/tests/domain/lfric/transformations/dynamo0p3_transformations_test.py:624]: Done.

File: src/psyclone/tests/domain/lfric/transformations/dynamo0p3_transformations_test.py [hunk 11]
```diff
@@ -945,13 +971,12 @@ def test_loop_fuse_different_spaces(monkeypatch, dist_mem):
...
-            # c and g halo exchange between loops can be moved before
-            # 1st loop as they are not accessed in first loop
-            schedule, _ = mtrans.apply(schedule.children[7],
-                                       schedule.children[6])
-            schedule, _ = mtrans.apply(schedule.children[8],
-                                       schedule.children[7])
-            index = 8
+            index = 9
+            # f, c and g halo exchanges between loops can be moved
+            # before 1st loop as they are not accessed in first loop
```
[the remark this answers is not in the document]
Comment (TeranIvy) [reply src/psyclone/tests/domain/lfric/transformations/dynamo0p3_transformations_test.py:976]: Done.

File: src/psyclone/tests/domain/lfric/transformations/dynamo0p3_transformations_test.py [hunk 12]
```diff
@@ -4734,41 +4778,65 @@ def test_rc_continuous_halo_remove():
...
-    assert "IF (f3_proxy%is_dirty(depth=1)) THEN" not in result
+    f3_inc_hex = schedule.children[2]
+    f3_inc_loop = schedule.children[4]
+    f3_read_hex = schedule.children[7]
+    f3_read_loop = schedule.children[9]
+    # f3 has "inc" access so there is a check for the halo exchange
+    # of depth 1, as well as two halo exchanges of depth 1 in code
+    # (one before f3_inc_loop and one before the f3_read_loop).
+    assert result.count("CALL f3_proxy%halo_exchange(depth=1") == 2
+    assert "IF (f3_proxy%is_dirty(depth=1)) THEN" in result
```
[the remark this answers is not in the document]
Comment (TeranIvy) [reply src/psyclone/tests/domain/lfric/transformations/dynamo0p3_transformations_test.py:4797]: Good idea, done!

File: src/psyclone/tests/dynamo0p3_test.py (hunk 3 above)
Comment (TeranIvy) [reply src/psyclone/tests/dynamo0p3_test.py:6663]: Logical thing to do that I forgot :smiley: Done.

File: src/psyclone/tests/domain/lfric/transformations/dynamo0p3_transformations_test.py [hunk 13]
```diff
@@ -4734,41 +4778,65 @@ def test_rc_continuous_halo_remove():
...
-    f3_write_loop = schedule.children[3]
-    f3_read_loop = schedule.children[7]
-    assert "CALL f3_proxy%halo_exchange(depth=1)" in result
-    assert "IF (f3_proxy%is_dirty(depth=1)) THEN" not in result
+    f3_inc_hex = schedule.children[2]
+    f3_inc_loop = schedule.children[4]
+    f3_read_hex = schedule.children[7]
+    f3_read_loop = schedule.children[9]
+    # f3 has "inc" access so there is a check for the halo exchange
+    # of depth 1, as well as two halo exchanges of depth 1 in code
```
[the remark this answers is not in the document]
Comment (TeranIvy) [reply src/psyclone/tests/domain/lfric/transformations/dynamo0p3_transformations_test.py:4794]: I meant "two" for the f3 field. Modified the text as sugested and added clarification.

File: src/psyclone/tests/domain/lfric/transformations/dynamo0p3_transformations_test.py [hunk 14]
```diff
@@ -4734,41 +4778,65 @@ def test_rc_continuous_halo_remove():
...
+    # (one before f3_inc_loop and one before the f3_read_loop).
```
[the remark this answers is not in the document]
Comment (TeranIvy) [reply src/psyclone/tests/domain/lfric/transformations/dynamo0p3_transformations_test.py:4795]: I meant "one" for the f3 field. There are actually three halo exchanges before the f3_read_loop due to access changes in two kernels (see the comment above).

**Comment by TeranIvy** [comment]
Addressed the review comments. Also wanted to update the branch with the master but @rupertford has already done that :smile: (thanks!) Tests with compilation pass. Ready for next review.

**Pull request merged**